Fixes #27687 docker logs --tails returns less lines than expected

shaynewang opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
Fixed issue when log file is too large and there is constant writing activity, the --tail command returns incorrect number of lines, and if the specified number larger than 100000 then the process hangs.

**- How I did it**
0. if the filesize is smaller than the buffer then load the whole file into memory goto 6
1.Changed the reading the last n lines algorithm to:
2.seek the starting position for a buffer size block from EOF
3.count the number of "\n"'s
4.if number of eol is larger than n then return
5.if not then double the size of the buffer then go back to 1.
6.take the last n line of buffered file and return

**- How to verify it**
docker logs --tail 10000 app | wc -l
should return number equal to the number of lines in a log file, or if the logfile has more lines then it should return 10000

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
for large log files tail now only read part of the file instead of the whole file

**- A picture of a cute animal (not mandatory but encouraged)**
![gp](http://cdn.attackofthecute.com/August-04-2011-04-29-47-twitter.jpeg)

The tailfile reader was loading the whole file into the memory before getting the last n files. It was using up a long of memory and have potential deadlock issue. Also when there is constant writing activity the tail returns wrong number of lines.

The fix uses buffer then let the seeker find the starting position of the buffer, if the buffer is too small then increase the size until the number of line are returned.

Signed-off-by: Shayne Wang <shaynexwang@gmail.com>

tonistiigi commented:
@shaynewang We shouldn't reread parts that have already been read for every buffer iteration. Instead use a constant size buffer and add them together or keep a list on buffers. Where is it reading the whole file in the memory in previous implementation, or are you reffering that this happens when new lines are being generated? Only problem I see with old implemention is that it uses `SEEK_END`.

shaynewang commented:
@tonistiigi You are correct I'll test the new fix and update my pull request.

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "27687-fix-tail-flag-returning-more-lines" git@github.com:shaynewang/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842354363528
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

shaynewang commented:
Passed tests
`# docker logs --tail 110 nostalgic_blackwell | wc -l
110
# docker logs --tail 10000000 nostalgic_blackwell|wc -l
85596
# docker logs --tail 100000 nostalgic_blackwell|wc -l

100000
# # docker logs --tail 10000 nostalgic_blackwell|wc -l
10000
# docker logs --tail 1 nostalgic_blackwell|wc -l
1
# docker logs --tail 0 nostalgic_blackwell|wc -l
0
# docker logs --tail -1 nostalgic_blackwell|wc -l

188090`

AkihiroSuda commented:
Can you add integration tests?

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "27687-fix-tail-flag-returning-more-lines" git@github.com:shaynewang/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842352617928
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

shaynewang commented:
@AkihiroSuda do you mean to write additional tests, or simply provide integration tests result?

AkihiroSuda commented:
Please write an automated test and put it under the integration-cli directory?

shaynewang commented:
@AkihiroSuda The integration tests are already implemented previously, adding one would be redundant. However, I did add a couple tests for edge cases. I'll add the commit once the all the tests are passed.

tonistiigi commented:
LGTM. Can you squash the commits to remove the previous implementation.

shaynewang committed:
Change reading order of tailfile

change reading order from beginning at the end to beginning at a buffer start
added intergration tests for boundary cases
Removed whitespace
Signed-off-by: Shayne Wang <shaynexwang@gmail.com>

vdemeester reviewed (approved):
LGTM 🐸
